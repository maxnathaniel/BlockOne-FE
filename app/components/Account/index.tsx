import React, { useState, useEffect, useCallback, memo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Dropdown, Input, Icon, Modal } from 'semantic-ui-react';
import numeral from 'numeral';

import { creditAccount } from '../../actions';
import {
  AccountContainer,
  AccountName,
  AccountType,
  AccountAmount,
  AccountNumber,
  BalanceAmount,
  BalanceLabel,
  Top,
  Left,
  Right,
  Error,
  Label,
  TransferMessage
} from './style';

interface IAccountProps {
  amount: string;
  name: string;
  index: number;
  id: string;
  type: string;
  isSelected: boolean;
  onHandleSelectAccount: (accountName: string) => void;
}

const Account: React.FC<IAccountProps> = memo(props => {
  const [selectedAccount, setSelectedAccount] = useState<string>('');
  const [showTransferModal, setShowTransferModal] = useState<boolean>(false);
  const [amount, setAmount] = useState<number>(0);
  const [nonNumericError, setNonNumericError] = useState<boolean>(false);
  const [zeroAmountError, setZeroAmountError] = useState<boolean>(false);
  const [exceededBalance, setExceededBalance] = useState<boolean>(false);
  const accountsList = Object.keys(useSelector((state: any) => state.accounts));
  const dispatch = useDispatch();

  const accountBalance = numeral(props.amount).format('$0,0.00');
  const nonNumericRegex = /[^0-9.]/;
  const zeroRegex = /^0+/;

  const onSelectAccount = useCallback((e: any, data: any) => {
    setSelectedAccount(data.value);
  }, []);

  const onTransferAmount = useCallback(() => {
    if (amount === 0 || !amount) {
      setZeroAmountError(true);
    } else {
      dispatch(creditAccount(props.name, selectedAccount, amount));
      setShowTransferModal(false);
    }
  }, [amount]);

  const onCloseTransactionsModal = useCallback(() => {
    setShowTransferModal(false);
    setNonNumericError(false);
    setZeroAmountError(false);
    setExceededBalance(false);
    setAmount(0);
  }, []);

  const onChange = useCallback((e: any) => {
    const amount = e.target.value;

    if (nonNumericRegex.test(amount)) {
      setNonNumericError(false);
    }

    if (zeroAmountError && amount !== 0) {
      setZeroAmountError(false);
    }

    if (exceededBalance && amount <= props.amount) {
      setExceededBalance(false);
    }

    if (nonNumericRegex.test(amount)) {
      setNonNumericError(true);
    } else if (zeroRegex.test(amount)) {
      setZeroAmountError(true);
    }

    if (amount > props.amount) {
      setExceededBalance(true);
    }

    setAmount(amount);
  }, [amount]);

  const onClick = useCallback((e: any) => {
    setShowTransferModal(true);
  }, []);

  const onClickAccount = useCallback((e: any) => {
    dispatch(props.onHandleSelectAccount(props.name));
  }, []);

  const accountOptions = accountsList.map(account => ({
    key: account, value: account, text: account })
  ).filter(account => account.value !== props.name);

  return (
    <AccountContainer index={props.index} isSelected={props.isSelected} onClick={onClickAccount}>
      <Modal size="mini" open={showTransferModal} onClose={onCloseTransactionsModal}>
        <Modal.Header>Transfer Money</Modal.Header>
        <Modal.Content>
          <BalanceAmount>{accountBalance}</BalanceAmount>
          <BalanceLabel>Balance</BalanceLabel>
          <Label>Recipent</Label>
          <Dropdown
            data-testid="select-account-dropdown"
            placeholder="Select Account"
            search
            selection
            fluid
            options={accountOptions}
            onChange={onSelectAccount}
          />
          <Label>Amount</Label>
          <Input
            data-testid="amount"
            transparent
            fluid
            type="text"
            onChange={onChange}
            placeholder="Amount..."
          />
          <Error data-testid="zero-amount-error">{zeroAmountError && 'Amount cannot be 0'}</Error>
          <Error data-testid="non-numeric-error">{nonNumericError && 'Amount must be a number'}</Error>
          <Error data-testid="exceeded-balance-error">{exceededBalance && 'Amount cannot be more than balance'}</Error>
          <TransferMessage>Transfering {numeral(amount).format('$0,0.00')} to {selectedAccount}</TransferMessage>
        </Modal.Content>
        <Modal.Actions>
          <Button
            data-testid="cancel-transfer-btn"
            color="red"
            size="medium"
            inverted
            onClick={onCloseTransactionsModal}
          >
            Cancel
          </Button>
          <Button
            data-testid="confirm-transfer-btn"
            color="teal"
            content="Confirm"
            size="medium"
            disabled={nonNumericError || zeroAmountError}
            onClick={onTransferAmount}
          />
        </Modal.Actions>
      </Modal>
      <Top>
        <AccountNumber>{props.id}</AccountNumber>
      </Top>
      <Left>
          <AccountName>
            {props.name}
        </AccountName>
        <AccountType>
          {props.type} account
        </AccountType>
      </Left>
      <Right>
        <AccountAmount>{accountBalance}</AccountAmount>
        <Icon data-testid="send-icon" name="send" onClick={onClick} />
      </Right>
    </AccountContainer>
  );
});

export default Account;