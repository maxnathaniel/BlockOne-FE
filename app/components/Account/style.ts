import styled from '../../styles/styled-components';

interface IAccountContainerProps {
  'min-width': string;
  'max-width': string;
  index: number;
  isSelected: boolean;
}

export const AccountContainer: any = styled.div<IAccountContainerProps>`
  display: block;
  color: #E0E3ED;
  cursor: pointer;
  border: ${props => props.isSelected ? '1px solid #E94F90' : ''};
  min-height: 90px;
  background-color: #272F4D;
  width: 95%;
  margin: 10px auto 0 auto;
  padding: 5px 20px;

  @media all and (max-width: 800px) {
    width: 90%;
  }
`;

export const AccountBody = styled.div`

`;

export const Top = styled.div`
  margin: 10px 0;
  text-align: left;
  width: 100%;
`;

export const Left = styled.div`
  display: inline-block;
  width: 50%;
`;

export const Right = styled.div`
  display: inline-block;
  text-align: right;
  width: 50%;
`;

export const AccountName = styled.h4`
  display: block;
  color: #66FCF1;
  margin: 10px 0 0 0;
`;

export const AccountType = styled.p`
  display: block;
  color: #E0E3ED;
  margin: 0;
`;

export const AccountAmount = styled.p`
  display: inline;
  margin: 5px 10px 0 0;
  text-align: right;
`;

export const AccountNumber = styled.p`
  margin: 5px 0 0 0;
`;

export const BalanceAmount = styled.h2`
  color: #66FCF1;
  margin: 0 0 0 0;
  text-align: center;
`;

export const BalanceLabel = styled.p`
  color: #FFF;
  font-size: 1em;
  text-align: center;
`;

export const MainAccountLabel = styled.span`
  padding: 2px 4px;
  max-width: 100px;
  margin-left: 20px;
  font-size: 0.8em;
  font-weight: 500;
  color: #40E0D0;
`;

export const AccountFooter = styled.div`
  border-top: 1px solid #dedede;
  color: #ababab;
  height: 10px;
`;

export const DropdownContainer = styled.div`
  display: inline-block;
`;

export const Error = styled.p`
  color: red;
  margin: 0;
`;

export const Label = styled.p`
  margin: 10px 0 5px 0;
`;

export const TransferMessage = styled.p`
  margin: 10px 0 0 0;
  color: #66FCF1;
`;