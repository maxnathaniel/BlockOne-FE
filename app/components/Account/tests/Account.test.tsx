import React from 'react';
import { render, fireEvent, queryByTestId, queryByText } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import numeral from 'numeral';
import configureMockStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

import Account from '../index';

const accountId = '13535hr-35afa2-25252fa-26396ahb';
const defaultState = {
  accounts: {
    'Block One Savings': { amount: 100000, id: accountId, type: 'Savings', isSelected: true, transactions: [] },
  }
};

const ZERO_ERROR_MSG = 'Amount cannot be 0';
const NON_NUMER_ERROR_MSG = 'Amount must be a number';
const EXCEEDED_BALANCE_ERROR_MSG = 'Amount cannot be more than balance';
// it's much easier to mock the react-redux APIs than to actually use the <Provider /> component
jest.mock('react-redux', () => ({
  useSelector: () => defaultState,
  useDispatch: () => jest.fn(),
}));

const mockStore = configureMockStore([createSagaMiddleware()]);
const store = mockStore(defaultState);

const amount = '10000';
const accountName = 'Test Account';
const accountBalance = numeral(amount).format('$0,0.00');
const onSelectAccountFn = jest.fn();

describe('<Account />', () => {

  it('should render the various inner Elements', () => {
    const { container, getByText }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    expect(getByText(accountName)).toBeInTheDocument();
    expect(getByText(accountId)).toBeInTheDocument();
    expect(getByText(accountBalance)).toBeInTheDocument();
    expect(getByText('Savings account')).toBeInTheDocument();

    const dropdown = queryByTestId(container, 'select-account-dropdown');
    const amountInput = queryByTestId(container, 'amount');
    const cancelTransferBtn = queryByTestId(container, 'cancel-transfer-btn');
    const confirmTransferBtn = queryByTestId(container, 'confirm-transfer-btn');

    expect(dropdown).not.toBeInTheDocument();
    expect(amountInput).not.toBeInTheDocument();
    expect(cancelTransferBtn).not.toBeInTheDocument();
    expect(confirmTransferBtn).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should open Modal on click', () => {
    const { container, getByTestId }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    const transferMoneyBtn = getByTestId('send-icon');
    fireEvent(transferMoneyBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));
    expect(getByTestId('select-account-dropdown')).toBeInTheDocument();
    expect(getByTestId('amount')).toBeInTheDocument();
    expect(getByTestId('cancel-transfer-btn')).toBeInTheDocument();
    expect(getByTestId('confirm-transfer-btn')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should close Modal on click of cancel btn', () => {
    const { container, getByTestId }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    const transferMoneyBtn = getByTestId('send-icon');
    fireEvent(transferMoneyBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));
    expect(getByTestId('select-account-dropdown')).toBeInTheDocument();
    expect(getByTestId('amount')).toBeInTheDocument();
    expect(getByTestId('cancel-transfer-btn')).toBeInTheDocument();
    expect(getByTestId('confirm-transfer-btn')).toBeInTheDocument();

    const cancelBtn = getByTestId('cancel-transfer-btn');
    fireEvent(cancelBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));

    const dropdown = queryByTestId(container, 'select-account-dropdown');
    const amountInput = queryByTestId(container, 'amount');
    const cancelTransferBtn = queryByTestId(container, 'cancel-transfer-btn');
    const confirmTransferBtn = queryByTestId(container, 'confirm-transfer-btn');

    expect(dropdown).not.toBeInTheDocument();
    expect(amountInput).not.toBeInTheDocument();
    expect(cancelTransferBtn).not.toBeInTheDocument();
    expect(confirmTransferBtn).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should display zero error if user inputs zero value in amount input field', () => {
    const { container, getByTestId, getByText }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    const transferMoneyBtn = getByTestId('send-icon');
    fireEvent(transferMoneyBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));
    expect(getByTestId('select-account-dropdown')).toBeInTheDocument();
    expect(getByTestId('amount')).toBeInTheDocument();
    expect(getByTestId('cancel-transfer-btn')).toBeInTheDocument();
    expect(getByTestId('confirm-transfer-btn')).toBeInTheDocument();

    // get a reference to the first child of element
    const inputField = getByTestId('amount').children[0];
    fireEvent.change(inputField, { target: { value: '0' } });

    expect(getByText(ZERO_ERROR_MSG)).toBeInTheDocument();

    const nonNumberErrorMsg = queryByText(container, NON_NUMER_ERROR_MSG);
    const exceededBalanceErrorMsg = queryByText(container, EXCEEDED_BALANCE_ERROR_MSG);

    expect(nonNumberErrorMsg).not.toBeInTheDocument();
    expect(exceededBalanceErrorMsg).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should display non numeric error if user inputs non numeric value in amount input field', () => {
    const { container, getByTestId, getByText }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    const transferMoneyBtn = getByTestId('send-icon');
    fireEvent(transferMoneyBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));
    expect(getByTestId('select-account-dropdown')).toBeInTheDocument();
    expect(getByTestId('amount')).toBeInTheDocument();
    expect(getByTestId('cancel-transfer-btn')).toBeInTheDocument();
    expect(getByTestId('confirm-transfer-btn')).toBeInTheDocument();

    // get a reference to the first child of element
    const inputField = getByTestId('amount').children[0];
    fireEvent.change(inputField, { target: { value: '#%@1' } });

    expect(getByText(NON_NUMER_ERROR_MSG)).toBeInTheDocument();

    const zeroErrorMsg = queryByText(container, ZERO_ERROR_MSG);
    const exceededBalanceErrorMsg = queryByText(container, EXCEEDED_BALANCE_ERROR_MSG);

    expect(zeroErrorMsg).not.toBeInTheDocument();
    expect(exceededBalanceErrorMsg).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should display exceed balance error if user inputs a value > balance in amount input field', () => {
    const { container, getByTestId, getByText }
      = render(
        <Account
          amount={amount}
          name={accountName}
          index={0}
          id={accountId}
          type={'Savings'}
          isSelected={true}
          onHandleSelectAccount={onSelectAccountFn}
        />
      );
    const transferMoneyBtn = getByTestId('send-icon');
    fireEvent(transferMoneyBtn,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }));
    expect(getByTestId('select-account-dropdown')).toBeInTheDocument();
    expect(getByTestId('amount')).toBeInTheDocument();
    expect(getByTestId('cancel-transfer-btn')).toBeInTheDocument();
    expect(getByTestId('confirm-transfer-btn')).toBeInTheDocument();

    // get a reference to the first child of element
    const inputField = getByTestId('amount').children[0];
    fireEvent.change(inputField, { target: { value: '110000' } });

    expect(getByText(EXCEEDED_BALANCE_ERROR_MSG)).toBeInTheDocument();

    const nonNumberErrorMsg = queryByText(container, NON_NUMER_ERROR_MSG);
    const zeroErrorMsg = queryByText(container, ZERO_ERROR_MSG);
    const exceededBalanceErrorMsg = queryByText(container, EXCEEDED_BALANCE_ERROR_MSG);

    expect(nonNumberErrorMsg).not.toBeInTheDocument();
    expect(zeroErrorMsg).not.toBeInTheDocument();
    expect(exceededBalanceErrorMsg).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});