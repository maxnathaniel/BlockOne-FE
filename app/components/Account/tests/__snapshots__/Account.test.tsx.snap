// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> should close Modal on click of cancel btn 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should display exceed balance error if user inputs a value > balance in amount input field 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should display non numeric error if user inputs non numeric value in amount input field 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should display zero error if user inputs zero value in amount input field 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should open Modal on click 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should render the various inner Elements 1`] = `
<div>
  <div
    class="sc-AykKC kHOAKO"
  >
    <div
      class="sc-AykKE lphdsI"
    >
      <p
        class="sc-AykKK ExoGy"
      >
        13535hr-35afa2-25252fa-26396ahb
      </p>
    </div>
    <div
      class="sc-AykKF dqtzlY"
    >
      <h4
        class="sc-AykKH gMPxuo"
      >
        Test Account
      </h4>
      <p
        class="sc-AykKI bEseiW"
      >
        Savings
         account
      </p>
    </div>
    <div
      class="sc-AykKG dnAkcX"
    >
      <p
        class="sc-AykKJ fWfswM"
      >
        $10,000.00
      </p>
      <i
        aria-hidden="true"
        class="send icon"
        data-testid="send-icon"
      />
    </div>
  </div>
</div>
`;
